tsl/src/nodes/decompress_chunk/exec.c

```diff
@@ -416,101 +442,94 @@ static TupleTableSlot *
+			 * save some CPU.
+			 */
+			ExprContext *econtext = state->csstate.ss.ps.ps_ExprContext;
+			ResetExprContext(econtext);
+
 			TupleTableSlot *subslot = ExecProcNode(linitial(state->csstate.custom_ps));
 
 			if (TupIsNull(subslot))
 				return NULL;
 
-			batch_done = false;
-			initialize_batch(state, subslot);
+			initialize_batch(state, subslot, slot);
 		}
 
-		ExecClearTuple(slot);
+		Assert(state->initialized);
+		Assert(state->total_batch_rows > 0);
+		Assert(state->current_batch_row < state->total_batch_rows);
 
-		for (i = 0; i < state->num_columns; i++)
+		for (int i = 0; i < state->num_columns; i++)
 		{
 			DecompressChunkColumnState *column = &state->columns[i];
-			switch (column->type)
+			if (column->type != COMPRESSED_COLUMN)
 			{
-				case COUNT_COLUMN:
-					if (state->counter <= 0)
-						/*
-						 * we continue checking other columns even if counter
-						 * reaches zero to sanity check all columns are in sync
-						 * and agree about batch end
-						 */
-						batch_done = true;
-					else
-						state->counter--;
-					break;
-				case COMPRESSED_COLUMN:
-				{
-					AttrNumber attr = AttrNumberGetAttrOffset(column->output_attno);
+				continue;
+			}
 
-					if (!column->compressed.iterator)
-					{
-						slot->tts_values[attr] = getmissingattr(slot->tts_tupleDescriptor,
-																attr + 1,
-																&slot->tts_isnull[attr]);
-					}
-					else
-					{
-						DecompressResult result;
-						result = column->compressed.iterator->try_next(column->compressed.iterator);
-
-						if (result.is_done)
-						{
-							batch_done = true;
-							continue;
-						}
-						else if (batch_done)
-						{
-							/*
-							 * since the count column is the first column batch_done
-							 * might be true if compressed column is out of sync with
-							 * the batch counter.
-							 */
-							elog(ERROR, "compressed column out of sync with batch counter");
-						}
-
-						slot->tts_values[attr] = result.val;
-						slot->tts_isnull[attr] = result.is_null;
-					}
+			const AttrNumber attr = AttrNumberGetAttrOffset(column->output_attno);
 
-					break;
-				}
-				case SEGMENTBY_COLUMN:
-				{
-					AttrNumber attr = AttrNumberGetAttrOffset(column->output_attno);
+			if (column->compressed.iterator != NULL)
+			{
+				DecompressResult result =
+					column->compressed.iterator->try_next(column->compressed.iterator);
 
-					slot->tts_values[attr] = column->segmentby.value;
-					slot->tts_isnull[attr] = column->segmentby.isnull;
-					break;
+				if (result.is_done)
+				{
+					elog(ERROR, "compressed column out of sync with batch counter");
 				}
-				case SEQUENCE_NUM_COLUMN:
-					/*
-					 * nothing to do here for sequence number
-					 * we only needed this for sorting in node below
-					 */
-					break;
+
+				slot->tts_isnull[attr] = result.is_null;
+				slot->tts_values[attr] = result.val;
 			}
 		}
 
-		if (batch_done)
+		/*
+		 * It's a virtual tuple slot, so no point in clearing/storing it
```

Comment:
It seems to work, but it might violate [the protocol](https://github.com/postgres/postgres/blob/a0137388cb642bcc1ba10fb906b88658653c589d/src/backend/executor/execTuples.c#L1544) for dealing with virtual tuples and we might miss [this free](https://github.com/postgres/postgres/blob/a0137388cb642bcc1ba10fb906b88658653c589d/src/backend/executor/execTuples.c#L109).

I wonder if that is intentional. If so, could you add an `Assert(TTS_IS_VIRTUAL(slot))` here?

Reply:
Could you add a comment that we are aware of this? Someone who does not know the discussion in this PR might otherwise wonder about it.

Reply:
Added a mention of the standard protocol.